File: 0238_product_of_array_except_self/product_except_self.go
package main

/**
 * Time: O(n)
 * Space: O(n)
 */
func productExceptSelf(nums []int) []int {
	size := len(nums)
	res := make([]int, size)
	for i := 0; i < size; i++ {
		res[i] = 1
	}
	pi, pj := 1, 1
	for i := 0; i < size; i++ {
		j := size - 1 - i
		res[i] *= pi
		res[j] *= pj
		pi *= nums[i]
		pj *= nums[j]
	}
	return res
}

func main() {

}

File: 0238_product_of_array_except_self/product_except_self.rb
def product_except_self(nums)
    # Time: O(n)
    # Space: O(n)
    n = nums.size
    res = [1] * n
    left = right = 1
    1.upto(n - 1) { |i|
        res[i] *= left *= nums[i - 1]
        res[~i] *= right *= nums[-i]
    }
    res
end